haystack/preview/components/reader.py

```diff
@@ -0,0 +1,239 @@
+        self.loaded = False
+        self.max_seq_length = max_seq_length
+        self.top_k = top_k
+        self.stride = stride
+        if os.path.exists(self.model):
+            return
+        try:
+            task = get_task(self.model)
+            if task != "question-answering":
+                raise ValueError(f"The provided model does not support question answering. Its intended use is {task}.")
+        except RuntimeError:
+            raise ValueError(
+                "The provided model either does not exist or it does not have a HF transformers compatible format."
+            )
+
+    def warm_up(self):
+        if not self.loaded:
+            if torch.cuda.is_available():
+                self.device = "cuda:0"
+            else:
+                self.device = "cpu:0"
+            self.model_ = AutoModelForQuestionAnswering.from_pretrained(self.model).to(self.device)
+            self.tokenizer = AutoTokenizer.from_pretrained(self.model)
+
+    def _flatten(self, queries: List[str], documents: List[List[Document]]) -> Tuple[List[str], List[str], List[int]]:
+        flattened_queries = [query for documents_, query in zip(documents, queries) for _ in documents_]
+        flattened_documents = [document.content for documents_ in documents for document in documents_]
+        query_ids = [i for i, documents_ in enumerate(documents) for _ in documents_]
+        return flattened_queries, flattened_documents, query_ids
+
+    def _preprocess(
+        self, queries: List[str], documents: List[str], max_seq_length: int, query_ids: List[int], stride: int
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, List[Encoding], List[int], List[int]]:
+        encodings = self.tokenizer(
+            queries,
+            documents,
+            padding=True,
+            truncation=True,
+            max_length=max_seq_length,
+            return_tensors="pt",
+            return_overflowing_tokens=True,
+            stride=stride,
+        )
+
+        input_ids = encodings.input_ids.to(self.device)
+        attention_mask = encodings.attention_mask.to(self.device)
+
+        query_ids = [query_ids[index] for index in encodings.overflow_to_sample_mapping]
+        document_ids = encodings.overflow_to_sample_mapping
+
+        encodings = encodings.encodings
+        sequence_ids = torch.tensor(
+            [[id_ if id_ is not None else -1 for id_ in encoding.sequence_ids] for encoding in encodings]
+        ).to(self.device)
+
+        return input_ids, attention_mask, sequence_ids, encodings, query_ids, document_ids
+
+    def _softmax_across_documents(self, logits: torch.Tensor, query_ids: List[int]) -> torch.Tensor:
+        last_docs = []
+        not_last_docs = []
+        i = 0
+        for j in range(query_ids[-1]):
+            while i < len(query_ids) and query_ids[i] == j:
+                if i != 0:
+                    not_last_docs.append(i - 1)
+                i += 1
+            last_docs.append(i - 1)
```

Comment:
Nitpick, let's change `i` and `j`. I know they're just used to group queries and docs but it helps a lot when reading if they have a more verbose name